main/src/ca/uwaterloo/flix/language/phase/unification/BoolMinimization.scala

```diff
@@ -0,0 +1,482 @@
+    */
+  def toNNF(f: Formula): Formula = {
+    import Formula._
+    f match {
+      case True => True
+
+      case False => False
+
+      case Var(v) => Var(v)
+
+      case not: Not => not.term match {
+        case True => False
+        case False => True
+        case Var(v) => mkNot(Var(v))
+        case Not(term) => term
+        case And(terms) => toNNF(mkOr(terms.map(t => mkNot(t))))
+        case Or(terms) => toNNF(mkAnd(terms.map(t => mkNot(t))))
+      }
+
+      case And(terms) => mkAnd(terms.map(toNNF))
+
+      case Or(terms) => mkOr(terms.map(toNNF))
+
+    }
+  }
+
+  /**
+    * Transform `f` to an equivalent formula that is a conjunction of
+    * disjunctions of either variables or negated variables.
+    * Ex. `x ∧ (x ∨ y) ∧ (z ∨ ¬y ∨ ¬z)`
+    */
+  def toCNF(f: Formula): Formula = {
+    def aux(f0: Formula): Formula = {
+      import Formula._
+      f0 match {
+        case True => True
+        case False => False
+        case Var(v) => Var(v)
+        case Not(Var(v)) => Not(Var(v))
+        case Not(_) => throw InternalCompilerException(s"Found complex negation after NNF ${show(f0)}")
+        case And(terms) => mkAnd(terms.map(toCNF))
+        case Or(terms) =>
+          // Note: this code relies on the ordering of simple terms before conjunctions
+          terms.foldLeft(False: Formula)(
+            (cnf, term) => term match {
+              case True => mkOr(cnf, True)
+              case False => mkOr(cnf, False)
+              case Var(v) => mkOr(cnf, Var(v))
+              case Not(Var(v)) => mkOr(cnf, Not(Var(v)))
+              case Not(_) => throw InternalCompilerException(s"Found complex negation after NNF ${show(term)}")
+              case And(terms0) => toCNF(mkAnd(terms0.map(t => mkOr(cnf, t))))
+              case Or(_) => throw InternalCompilerException(s"Found directly nested Or: ${show(f0)}")
+            })
+      }
+    }
+
+    aux(toNNF(f))
+  }
+
+  /**
+    * Transform `f` to an equivalent formula that is a disjunction of
+    * conjunctions of either variables or negated variables.
+    * Ex. `x ∨ (x ∧ y) ∨ (z ∧ ¬y ∧ ¬z)`
+    */
+  def toDNF(f: Formula): Formula = ???
+
+  def fromType(t: Type): Formula = {
+    import Formula._
+    t match {
+      case Type.True => True
+      case Type.False => False
+      case v: Type.KindedVar => Var(v)
+      case TypeNot(t0) => mkNot(fromType(t0))
+      case TypeAnd(t1, t2) => mkAnd(fromType(t1), fromType(t2))
+      case TypeOr(t1, t2) => mkOr(fromType(t1), fromType(t2))
+      case other => throw InternalCompilerException(s"Unexpected non-boolean type ${other.toString}")
+    }
+  }
+
+  def toType(f: Formula): Type = {
+    import Formula._
+    f match {
+      case True => Type.True
+      case False => Type.False
+      case Var(v) => v
+      case Not(term) => TypeNot(toType(term))
+      case And(terms) => terms match {
+        case Nil => throw InternalCompilerException("Cannot transform empty conjunction")
```

Comment:
I got this exception a few times when playing around.

Reply:
Yeah it feels a bit brittle, I should make the data types more constrained. It would also be nice to structurally disallow directly nested ands/ors